python/benchmark/gen_data_distributed.py

```diff
@@ -232,6 +234,152 @@ def make_matrix_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
+        if num_partitions is None:
+            num_partitions = spark.sparkContext.defaultParallelism
+
+        # Retrieve input params or set to defaults.
+        seed = params["random_state"]
+        generator = np.random.RandomState(seed)
+        bias = params.get("bias", 0.0)
+        noise = params.get("noise", 0.0)
+        shuffle = params.get("shuffle", True)
+        effective_rank = params.get("effective_rank", None)
+        n_informative = params.get("n_informative", 10)
+        n_targets = params.get("n_targets", 1)
+
+        # Input set is either well conditioned (default) or has a low rank fat tail singular profile (see LowRankMatrixDataGen).
+        # Output is generated by applying a (potentially biased) random linear regression model to the input, with n_informative
+        # nonzero regressors and some gaussian centered noise with adjustable scale.
+        # Code adapted from sklearn.datasets.make_regression().
+        if effective_rank is not None:
+            tail_strength = params.get("tail_strength", 0.5)
+            lrm_input_args = [
+                "--num_rows",
+                str(rows),
+                "--num_cols",
+                str(cols),
+                "--dtype",
+                str(dtype),
+                "--output_dir",
+                "temp",
+                "--output_num_files",
+                str(num_partitions),
+                "--effective_rank",
+                str(effective_rank),
+                "--tail_strength",
+                str(tail_strength),
+                "--random_state",
+                str(seed),
+            ]
+            # Generate a low-rank, fat tail input set.
+            X, _ = LowRankMatrixDataGen(lrm_input_args).gen_dataframe(spark)
+            assert X.rdd.getNumPartitions() == num_partitions, (
+                f"Unexpected num partitions received from LowRankMatrix;"
+                f"expected {num_partitions}, got {X.rdd.getNumPartitions()}"
+            )
+        else:
+            # Randomly generate a well-conditioned input set.
+            X = spark.createDataFrame(
+                RandomRDDs.normalVectorRDD(
+                    spark.sparkContext,
+                    rows,
+                    cols,
+                    numPartitions=num_partitions,
+                    seed=seed,
+                ).map(lambda nparray: nparray.tolist()),
+                schema=",".join(self.schema),
+            )
+
+            assert X.rdd.getNumPartitions() == num_partitions, (
+                f"Unexpected num partitions received from RandomRDDs;"
+                f"expected {num_partitions}, got {X.rdd.getNumPartitions()}"
+            )
+
+        # Generate ground truth upfront.
+        ground_truth = np.zeros((cols, n_targets))
+        ground_truth[:n_informative, :] = 100 * generator.uniform(
+            size=(n_informative, n_targets)
+        )
+
+        if shuffle:
```

Comment:
Why shuffle is required in this class?

Reply:
Low rank matrix / randomRDDs don't have any shuffling of their own, so row-wise shuffle per partition is meant to reduce some of the potential sample generation bias. Column-wise shuffle is to prevent same indices to be set as informative features every time.